src/Mvc/Mvc.ViewFeatures/test/Filters/AntiforgeryApplicationModelProviderTest.cs

```diff
@@ -0,0 +1,199 @@
+
+        // Assert
+        var controller = Assert.Single(context.Result.Controllers);
+        if (filterType is not null)
+        {
+            var filter = Assert.Single(controller.Filters);
+            Assert.IsType(filterType, filter);
+        }
+        else
+        {
+            Assert.Empty(controller.Filters);
+        }
+    }
+
+    [Theory]
+    [InlineData(typeof(DerivedAntiforgeryMetadataController), typeof(AntiforgeryMiddlewareAuthorizationFilter))]
+    [InlineData(typeof(DerivedAntiforgeryMetadataMvcAttributeController), typeof(ValidateAntiForgeryTokenAttribute))]
+    [InlineData(typeof(DerivedMvcAttributeController), typeof(ValidateAntiForgeryTokenAttribute))]
+    [InlineData(typeof(DerivedEmptyController), null)]
+    public void WorksWithAttributesOnDerivedController(Type controllerType, Type filterType)
+    {
+        var provider = new AntiforgeryApplicationModelProvider(
+            Options.Create(new MvcOptions()),
+            NullLogger<AntiforgeryMiddlewareAuthorizationFilter>.Instance);
+        var context = CreateProviderContext(controllerType);
+
+        // Act
+        provider.OnProvidersExecuting(context);
+
+        // Assert
+        var controller = Assert.Single(context.Result.Controllers);
+        if (filterType is not null)
+        {
+            var filter = Assert.Single(controller.Filters);
+            Assert.IsType(filterType, filter);
+        }
+        else
+        {
+            Assert.Empty(controller.Filters);
+        }
+    }
+
+    private static ApplicationModelProviderContext CreateProviderContext(Type controllerType)
+    {
+        var defaultProvider = new DefaultApplicationModelProvider(
+            Options.Create(new MvcOptions()),
+            new EmptyModelMetadataProvider());
+
+        var context = new ApplicationModelProviderContext(new[]
+        {
+            controllerType.GetTypeInfo()
+        });
+        defaultProvider.OnProvidersExecuting(context);
+        return context;
+    }
+
+    private static ActionModel GetActionModel(
+        string actionName,
+        object[] actionAttributes = null,
+        object[] controllerAttributes = null)
+    {
+        actionAttributes ??= Array.Empty<object>();
+        controllerAttributes ??= Array.Empty<object>();
+
+        var controllerModel = new ControllerModel(typeof(TestController).GetTypeInfo(), controllerAttributes);
+        var actionModel = new ActionModel(typeof(TestController).GetMethod(actionName), actionAttributes)
+        {
+            Controller = controllerModel,
+        };
+
+        controllerModel.Actions.Add(actionModel);
+
+        return actionModel;
+    }
+
+    private class TestController
+    {
+        [HttpPost("with-antiforgery-metadata")]
+        [RequireAntiforgeryToken]
+        public IActionResult WithAntiforgeryMetadata() => null;
+
+        [HttpPost("with-antiforgery-metadata-mvc-attribute")]
+        [RequireAntiforgeryToken]
+        [ValidateAntiForgeryToken]
```

Comment:
Is there a chance we can just throw if we see the two different attributes applied? It will make our lives easier.

Reply:
Yep, we can do this.